src/ORM/Rule/ExistsIn.php

```diff
@@ -125,6 +126,51 @@ public function __invoke(EntityInterface $entity, array $options)
             }
         }
 
+        if (isset($options['parent']) && is_a($options['parent'], EntityInterface::class)) {
+            /**
+             * @var EntityInterface $parent
+             */
+            $parent = $options['parent'];
+
+            //extract entity from given parent regarding HM and HABTM associations
+            $parentEntity = null;
+            if ($isAssociation) {
+                if (TableRegistry::get($parent->getSource()) === $target->getTarget()) {
+                    //parent is of same  as target
+                    $parentEntity = $parent;
+                } else {
+                    if (!empty($parent[strtolower($target->getName())])) {
+                        $parentEntity = $parent[strtolower($target->getName())];
+                    }
+                }
+            } else {
+                $parentEntity = $parent;
+            }
```

Comment:
These conditionals have quite a bit of nesting, would it be possible to move some of that complexity into a protected method that can return early as needed? That would let you have less indentation/complexity and still be able to find the parent entity.